File: Y2/POO/lab11/composite_ex.cpp
#include <iostream>
#include <string>
#include <vector>


using namespace std;

class Component{
    public:
    virtual void execute() = 0;
    virtual void add(Component* c) {};
    ~Component(){}
};

class UI: public Component{
    vector<Component*> components;

    public:
    void add(Component* c){
        components.push_back(c);
    }

    void execute(){
        for(Component* c : components){
            c->execute();
        }
    }
};

class Button: public Component{
    public:
    void execute(){
        cout<<"Button pressed.\n";
    }
};

class Label: public Component{
    string text;
    public:
    Label(const string& _text): text(_text){}

    void execute(){
        cout<<"Label text is: " + text + "\n";
    }
};

class Window: public Component{
    vector<Component*> components;

    public:
    void add(Component* c){
        components.push_back(c);
    }

    void execute(){
        for(Component* c : components){
            c->execute();
        }
    }
};




int main(){
    UI *ui = new UI();

    Window *window = new Window();

    ui->add(window);
    
    window->add(new Button());
    window->add(new Label("Ce mai faci"));

    ui->execute();
}

File: Y2/POO/lab11/kotlin/composite_ex.kt
abstract class Component {
    open fun add(c: Component) {}
    abstract fun execute()
}

class UI : Component() {
    private val components = mutableListOf<Component>()

    override fun add(c: Component) {
        components.add(c)
    }

    override fun execute() {
        for (c in components) {
            c.execute()
        }
    }
}

class Button : Component() {
    override fun execute() {
        println("Button pressed.")
    }
}

class Label(private val text: String) : Component() {
    override fun execute() {
        println("Label text is: $text")
    }
}

class Window : Component() {
    private val components = mutableListOf<Component>()

    override fun add(c: Component) {
        components.add(c)
    }

    override fun execute() {
        for (c in components) {
            c.execute()
        }
    }
}

fun main() {
    val ui = UI()
    val window = Window()

    ui.add(window)

    window.add(Button())
    window.add(Label("Ce mai faci"))

    ui.execute()
}
